### PR TITLE
gsm soft reset

### PR DESCRIPTION
**Pull request by SinaBaghi**
Implementing modem soft reset and dynamic APN set at setup stage

### CODE REVIEW

File: drivers/modem/gsm_ppp.c [hunk 1]
```diff
@@ -1172,13 +1229,32 @@ void gsm_ppp_stop(const struct device *dev)
 	net_if_l2(iface)->enable(iface, false);
 
 	if (IS_ENABLED(CONFIG_GSM_MUX)) {
+		gsm_soft_reset(gsm);
```
Comment (SinaBaghi) [inline drivers/modem/gsm_ppp.c:1242]: If the mux is enabled we have AT channel access we can perform modem reset directly. After the soft reset muxing on modem side side is disabled and we disable muxing here too.

File: drivers/modem/gsm_ppp.c [hunk 2]
```diff
@@ -1172,13 +1229,32 @@ void gsm_ppp_stop(const struct device *dev)
...
 		/* Lower mux_enabled flag to trigger re-sending AT+CMUX etc */
 		gsm->mux_enabled = false;
 
 		if (gsm->ppp_dev) {
 			uart_mux_disable(gsm->ppp_dev);
 		}
 	}
+	else {
+		if (gsm->setup_done) {
```
Comment (SinaBaghi) [inline drivers/modem/gsm_ppp.c:1251]: If the setup is done if the muxing is not enabled, modem would be in data mode and AT commands won't be accessible. First we need to exit PPP mode and then execute soft reset using AT channel.

File: drivers/modem/gsm_ppp.c [hunk 3]
```diff
@@ -1160,6 +1178,47 @@ void gsm_ppp_start(const struct device *dev)
...
+			retry--;
+			int ret = modem_cmd_send_nolock(&gsm->context.iface,
+									&gsm->context.cmd_handler,
+									&response_cmds[0],
+				    			ARRAY_SIZE(response_cmds),
+									"AT+CFUN=0",
+									&gsm->sem_response, K_SECONDS(2));
+			if(ret == 0) {
+				break;
+			}
```
Comment (MrOneTwo) [inline drivers/modem/gsm_ppp.c:1197]: we'll fail silently here. Should we add a print or not worth it?

**Review by MrOneTwo — APPROVED** [review]

File: drivers/modem/gsm_ppp.c (hunk 3 above)
Comment (SinaBaghi) [reply drivers/modem/gsm_ppp.c:1197]: In some network loss events modem goes to some unknown states and won't answer to AT commands and trying more is just a safety guard but if doesn't then there is nothing we can do other than power cycling modem. I think it's good enough. But I can add an error log here

**Pull request merged**